File: packages/kits/AriaMcpKit/Sources/AriaMCP/ModeSessionState.swift
/// Per-session mode sticky state and call counters for the modes coaching system.
///
/// ## Lifecycle
///
/// One `ModeSessionState` instance is created per `ToolDispatcher`. For the
/// stdio transport (one process per session), this gives per-session stickiness
/// for the session's lifetime. For the HTTP transport (resident process, multiple
/// clients), the dispatcher is shared; the session state is therefore shared too.
///
/// ## Extension point: per-client-id sticky state (HTTP)
///
/// The spec calls for per-client-id mode state with a TTL on the HTTP path. That
/// requires a per-client-id map keyed by the client's session ID, which the HTTP
/// server negotiates during `initialize`. The current implementation uses a single
/// shared instance — correct for stdio (one client per process) and advisory-
/// accurate for HTTP (shared state still applies the last-declared wins rule;
/// two clients may temporarily see each other's mode, which is harmless because
/// modes are advisory). A future HTTP session manager would inject a per-client
/// `ModeSessionState` by wrapping `ToolDispatcher` per request, following the
/// same pattern as `SensitivityGrantLedger`.
///
/// ## Preference keys
///
/// The spec defines two preference keys read from the estate manifest:
///   `modes.sticky_enabled` (default true)  — when false, mode declarations
///       are accepted and a hint is returned but the sticky state is never set.
///       Advisory-only mode is the result, same as absent sticky state. One code
///       path, two reasons (off by preference vs off by absence).
///   `modes.coaching_calls` (default 25, 0=off) — how often the coaching block
///       fires. The estate manifest equivalent of an explicit per-call counter
///       override.
///
/// Both preferences are read at session start via
/// `GeniusLocusKit.provisionedModesConfig(for:)` and applied by
/// `ToolDispatcher.dispatch` on the first tool call. `applyPreferences` is the
/// seam; `setCoachingCallsX` remains for tests that override the cadence
/// per-call without going through the estate manifest.
///
/// ## Bool and bitmap discipline
///
/// No Bool stored properties on session state. The `stickyEnabled` preference
/// is a computed accessor over bit 0 of `sessionPreferenceBitmap` (Int64) per
/// the fleet-wide bitmap rule. Bit 1 (`configuredFromEstate`) guards the
/// apply-once contract. This applies to non-persisted actors the same as
/// database entities.
///
/// ## Sendable conformance
///
/// `ModeSessionState` is an `actor`, making it `Sendable`. It is held as `let`
/// in the `Sendable` struct `ToolDispatcher`, satisfying the concurrency checker.

import Foundation

// MARK: - CoachingSnapshot

/// Immutable snapshot of per-session call counters, passed to `PeriodicCoach`
/// for deterministic block rendering.
public struct CoachingSnapshot: Sendable, Equatable {
    /// Total moot tool calls this session.
    public let totalCalls: Int
    /// Calls per tool name (canonical tool name → count).
    public let toolCounts: [String: Int]
    /// Bigram counts: "toolA→toolB" → how many times toolB immediately followed toolA.
    public let bigramCounts: [String: Int]
    /// Mode attribution counts: mode name → calls attributed to that mode.
    public let modeAttributionCounts: [String: Int]
}

// MARK: - ModeSessionState

/// Actor carrying sticky mode declaration and per-session call counters.
///
/// All mutations are async and actor-isolated — no external synchronization needed.
public actor ModeSessionState: Sendable {

    // MARK: - Preference bitmap
    // Bit assignments for sessionPreferenceBitmap (Int64):
    //   bit 0 — stickyEnabled         (default 1 = true): sticky state is recorded
    //   bit 1 — configuredFromEstate  (default 0 = false): applyPreferences has run
    //   bits 2-63 — reserved

    /// Preference flags for this session (bitmap per fleet-wide house-style rule).
    /// Default value 0b01 = stickyEnabled is on; configuredFromEstate starts false.
    private var sessionPreferenceBitmap: Int64 = 0b01

    /// When false, mode declarations are accepted and may produce hints, but the
    /// sticky state is never updated — same advisory-only code path as absent state.
    /// Default per spec: true. The estate manifest key is `modes.sticky_enabled`.
    public var stickyEnabled: Bool {
        get { sessionPreferenceBitmap & (1 << 0) != 0 }
        set {
            if newValue { sessionPreferenceBitmap |= (1 << 0) }
            else { sessionPreferenceBitmap &= ~(1 << 0) }
        }
    }

    /// Whether `applyPreferences` has been called at least once this session.
    ///
    /// Guards the apply-once contract in `ToolDispatcher.dispatch`: the first
    /// tool call reads the estate manifest and calls `applyPreferences`; all
    /// subsequent calls skip the read (the manifest is RAM-resident so the
    /// overhead is low, but idempotency is cleaner and avoids unnecessary
    /// async work on every call).
    ///
    /// Backed by bitmap bit 1 per fleet-wide house-style rule (no Bool stored
    /// properties, even on non-persisted session actors).
    var configuredFromEstate: Bool {
        get { sessionPreferenceBitmap & (1 << 1) != 0 }
        set {
            if newValue { sessionPreferenceBitmap |= (1 << 1) }
            else { sessionPreferenceBitmap &= ~(1 << 1) }
        }
    }

    /// How many moot tool calls between coaching blocks. 0 = off.
    /// Default per spec: 25. The estate manifest key is `modes.coaching_calls`.
    public var coachingCallsX: Int = 25

    // MARK: - Sticky state

    /// The last-declared mode declaration, retained as the session default.
    ///
    /// A bare mode name (no variant) clears any prior variant for that mode but
    /// keeps the mode name for attribution. A declaration for a *different* mode
    /// replaces the previous declaration entirely.
    ///
    /// Never set when `stickyEnabled == false`.
    private(set) var stickyDeclaration: ModeDeclaration? = nil

    // MARK: - Sticky recall answer mode

    /// The answer mode raw value for the current sticky Recall variant, or `nil`
    /// when no sticky Recall=<variant> is set.
    ///
    /// Injected as the `answer` argument by the pre-decode transform hook when
    /// `answer` is absent from the call arguments and the session has a sticky
    /// Recall variant. Per-call explicit `answer` always wins — the transform hook
    /// only injects when the argument is absent.
    ///
    /// Returns `"auto"` for `Recall=Auto`, `"never"` for `Recall=Rows`,
    /// `"always"` for `Recall=Answer`.
    public var stickyRecallAnswerMode: String? {
        stickyDeclaration?.recognizedRecallVariant?.answerModeRawValue
    }

    // MARK: - Call counters

    /// Total calls recorded this session.
    private(set) var totalCallCount: Int = 0

    /// Calls per tool name.
    private(set) var toolCallCounts: [String: Int] = [:]

    /// The tool name dispatched in the immediately preceding call (for bigram tracking).
    private var lastToolName: String? = nil

    /// Next-call bigram frequency: "toolA→toolB" → count.
    private(set) var bigramCounts: [String: Int] = [:]

    /// Mode attribution: mode name → calls attributed to that mode.
    /// Attribution uses the declared mode when present, otherwise infers from the
    /// tool name via `MootMode.inferredBundle(for:)`.
    private(set) var modeAttributionCounts: [String: Int] = [:]

    // MARK: - Init

    public init() {}

    // MARK: - Test seam

    /// Override the coaching cadence. Used by tests to trigger coaching quickly
    /// without going through the estate manifest. Sets `configuredFromEstate = true`
    /// so that `ToolDispatcher.dispatch` skips the provisioned-config read on the
    /// first tool call — the seam takes full precedence over the estate key.
    func setCoachingCallsX(_ value: Int) {
        coachingCallsX = value
        configuredFromEstate = true
    }

    // MARK: - Estate provisioned preferences

    /// Apply preferences read from the estate manifest (via
    /// `GeniusLocusKit.provisionedModesConfig(for:)`).
    ///
    /// Called by `ToolDispatcher.dispatch` on the first tool call of the session.
    /// Subsequent calls are no-ops (guarded by `configuredFromEstate`).
    ///
    /// - Parameters:
    ///   - stickyEnabled: from `ModesManifest.stickyEnabled`. `false` makes all
    ///     mode declarations advisory-only (no sticky state recorded).
    ///   - coachingCalls: from `ModesManifest.coachingCalls`. `0` suppresses all
    ///     coaching blocks. Replaces the spec-default `coachingCallsX = 25`.
    func applyPreferences(stickyEnabled: Bool, coachingCalls: Int) {
        guard !configuredFromEstate else { return }
        self.stickyEnabled = stickyEnabled
        self.coachingCallsX = coachingCalls
        configuredFromEstate = true
    }

    // MARK: - Recording a call

    /// Record a completed tool call and return the new total call count.
    ///
    /// - Parameters:
    ///   - toolName: The MCP tool name that was dispatched.
    ///   - mode: The parsed mode declaration from the `mode` arg, or nil when absent.
    /// - Returns: The updated `totalCallCount` after recording this call.
    @discardableResult
    public func recordCall(toolName: String, mode: ModeDeclaration?) -> Int {
        // Update sticky when a mode arg is present, sticky is enabled, AND the mode
        // name is recognized. An unrecognized mode declaration is IGNORED ENTIRELY for
        // sticky purposes (ruling W4): it must not clobber an existing valid sticky
        // declaration. The AI is still notified via a hint line (handled by the
        // dispatch layer), so the fail-open contract is preserved — only sticky state
        // is unaffected. This matches the Rust `record_call` implementation.
        if stickyEnabled, let m = mode, m.recognizedMode != nil {
            stickyDeclaration = m
        }

        // Increment total and per-tool counters.
        totalCallCount += 1
        toolCallCounts[toolName, default: 0] += 1

        // Bigram: record toolA→toolB pair from the previous call.
        if let last = lastToolName {
            let bigram = "\(last)→\(toolName)"
            bigramCounts[bigram, default: 0] += 1
        }
        lastToolName = toolName

        // Mode attribution: declared mode name beats inferred bundle.
        let attributedMode: String?
        if let m = mode, m.recognizedMode != nil {
            attributedMode = m.modeName
        } else {
            attributedMode = MootMode.inferredBundle(for: toolName)?.rawValue
        }
        if let m = attributedMode {
            modeAttributionCounts[m, default: 0] += 1
        }

        return totalCallCount
    }

    // MARK: - Coaching trigger

    /// Returns true when the current call count is an exact multiple of
    /// `coachingCallsX`. Returns false when `coachingCallsX` is 0 (off).
    ///
    /// Must be called AFTER `recordCall` so `totalCallCount` reflects the current call.
    public func shouldCoach() -> Bool {
        guard coachingCallsX > 0 else { return false }
        return totalCallCount % coachingCallsX == 0
    }

    // MARK: - Snapshot

    /// Return an immutable snapshot for deterministic block rendering by `PeriodicCoach`.
    public var snapshot: CoachingSnapshot {
        CoachingSnapshot(
            totalCalls: totalCallCount,
            toolCounts: toolCallCounts,
            bigramCounts: bigramCounts,
            modeAttributionCounts: modeAttributionCounts
        )
    }
}

File: packages/kits/AriaMcpKit/rust/src/mode_session_state.rs
//! Per-session mode sticky state and call counters for the modes coaching system.
//!
//! ## Thread safety
//!
//! `ModeSessionState` uses `std::sync::Mutex` for interior mutability, matching
//! the `Dispatcher`'s `&self` interface requirement. The Mutex wraps the inner
//! mutable state; the outer `ModeSessionState` is `Send + Sync`.
//!
//! ## Preference bitmap
//!
//! Bit 0 of `session_preference_bitmap` (i64) = `sticky_enabled` (default 1 = true).
//! No Bool stored fields — bitmap pattern per fleet-wide house-style rule.
//!
//! ## Extension points
//!
//! - Per-client-id sticky state for HTTP: the current implementation uses one
//!   shared state per `Dispatcher` instance (correct for stdio, advisory for HTTP).
//!   HTTP extension: wrap `Dispatcher` per-request and inject a fresh `ModeSessionState`.
//! - Preference keys from the estate manifest: `modes.sticky_enabled` and
//!   `modes.coaching_calls` are read at session start via
//!   `EstateCoordinator::provisioned_modes_config` and applied by
//!   `Dispatcher::tools_call` through `ModeSessionState::apply_preferences`.
//!
//! Parity: Rust twin of Swift `ModeSessionState.swift`.

use std::collections::HashMap;
use std::sync::Mutex;

use crate::mode_registry::ModeDeclaration;

// MARK: - CoachingSnapshot

/// Immutable snapshot of per-session call counters, passed to `PeriodicCoach`
/// for deterministic block rendering.
#[derive(Debug, Clone, PartialEq)]
pub struct CoachingSnapshot {
    /// Total moot tool calls this session.
    pub total_calls: usize,
    /// Calls per tool name.
    pub tool_counts: HashMap<String, usize>,
    /// Bigram counts: "toolA→toolB" → count.
    pub bigram_counts: HashMap<String, usize>,
    /// Mode attribution counts: mode name → calls attributed to that mode.
    pub mode_attribution_counts: HashMap<String, usize>,
}

// MARK: - Inner mutable state

struct ModeSessionStateInner {
    /// Preference bitmap — bit assignments:
    ///   bit 0 = sticky_enabled        (default 1 = true)
    ///   bit 1 = configured_from_estate (default 0 = false)
    ///   bits 2-63 = reserved
    /// No Bool stored fields — bitmap pattern per fleet-wide house-style rule.
    session_preference_bitmap: i64,
    /// How many moot tool calls between coaching blocks. 0 = off. Default: 25.
    coaching_calls_x: usize,
    /// The last-declared mode, or None.
    sticky_declaration: Option<ModeDeclaration>,
    /// Total calls this session.
    total_call_count: usize,
    /// Calls per tool name.
    tool_call_counts: HashMap<String, usize>,
    /// The previous call's tool name (for bigram tracking).
    last_tool_name: Option<String>,
    /// Bigram frequency map.
    bigram_counts: HashMap<String, usize>,
    /// Mode attribution counts.
    mode_attribution_counts: HashMap<String, usize>,
}

impl ModeSessionStateInner {
    fn new() -> Self {
        ModeSessionStateInner {
            // bit 0 = sticky_enabled = true; bit 1 = configured_from_estate = false
            session_preference_bitmap: 0b01,
            coaching_calls_x: 25,
            sticky_declaration: None,
            total_call_count: 0,
            tool_call_counts: HashMap::new(),
            last_tool_name: None,
            bigram_counts: HashMap::new(),
            mode_attribution_counts: HashMap::new(),
        }
    }

    fn sticky_enabled(&self) -> bool {
        self.session_preference_bitmap & (1 << 0) != 0
    }

    fn configured_from_estate(&self) -> bool {
        self.session_preference_bitmap & (1 << 1) != 0
    }

    fn set_sticky_enabled(&mut self, value: bool) {
        if value {
            self.session_preference_bitmap |= 1 << 0;
        } else {
            self.session_preference_bitmap &= !(1 << 0);
        }
    }

    fn set_configured_from_estate(&mut self, value: bool) {
        if value {
            self.session_preference_bitmap |= 1 << 1;
        } else {
            self.session_preference_bitmap &= !(1 << 1);
        }
    }
}

// MARK: - ModeSessionState

/// Session-scoped mode sticky state and call counters.
///
/// Thread-safe via `Mutex`. Create one instance per `Dispatcher`; for HTTP
/// where per-client isolation is required, inject a fresh instance per request.
pub struct ModeSessionState {
    inner: Mutex<ModeSessionStateInner>,
}

impl ModeSessionState {
    /// Construct with spec defaults (sticky_enabled = true, coaching_calls_x = 25).
    pub fn new() -> Self {
        ModeSessionState {
            inner: Mutex::new(ModeSessionStateInner::new()),
        }
    }

    /// Record a completed tool call and return the new total call count.
    ///
    /// Updates sticky declaration (when sticky_enabled), increments counters,
    /// records bigram from previous call, and updates mode attribution.
    pub fn record_call(&self, tool_name: &str, mode: Option<&ModeDeclaration>) -> usize {
        let mut inner = self.inner.lock().expect("ModeSessionState lock poisoned");

        // Update sticky declaration when sticky is enabled, a mode was declared, AND
        // the mode name is recognized. An unrecognized mode declaration is IGNORED
        // ENTIRELY for sticky purposes (ruling W4): it must not clobber an existing
        // valid sticky declaration. The AI is notified via a hint line by the dispatcher,
        // so fail-open is preserved — only sticky state is unaffected.
        if inner.sticky_enabled() {
            if let Some(m) = mode {
                if m.recognized_mode().is_some() {
                    inner.sticky_declaration = Some(m.clone());
                }
            }
        }

        // Increment total and per-tool counters.
        inner.total_call_count += 1;
        *inner.tool_call_counts.entry(tool_name.to_string()).or_insert(0) += 1;

        // Bigram: record toolA→toolB pair from the previous call.
        if let Some(last) = inner.last_tool_name.clone() {
            let bigram = format!("{}→{}", last, tool_name);
            *inner.bigram_counts.entry(bigram).or_insert(0) += 1;
        }
        inner.last_tool_name = Some(tool_name.to_string());

        // Mode attribution: declared mode name beats inferred bundle.
        let attributed_mode = if let Some(m) = mode {
            if m.recognized_mode().is_some() {
                Some(m.mode_name.clone())
            } else {
                crate::mode_registry::MootMode::inferred_bundle(tool_name)
                    .map(|mm| mm.raw_value().to_string())
            }
        } else {
            crate::mode_registry::MootMode::inferred_bundle(tool_name)
                .map(|mm| mm.raw_value().to_string())
        };
        if let Some(m) = attributed_mode {
            *inner.mode_attribution_counts.entry(m).or_insert(0) += 1;
        }

        inner.total_call_count
    }

    /// Returns true when coaching should fire for this call count.
    ///
    /// Must be called AFTER `record_call` so total_call_count reflects the current call.
    pub fn should_coach(&self) -> bool {
        let inner = self.inner.lock().expect("ModeSessionState lock poisoned");
        if inner.coaching_calls_x == 0 {
            return false;
        }
        inner.total_call_count % inner.coaching_calls_x == 0
    }

    /// Return an immutable snapshot for deterministic block rendering.
    pub fn snapshot(&self) -> CoachingSnapshot {
        let inner = self.inner.lock().expect("ModeSessionState lock poisoned");
        CoachingSnapshot {
            total_calls: inner.total_call_count,
            tool_counts: inner.tool_call_counts.clone(),
            bigram_counts: inner.bigram_counts.clone(),
            mode_attribution_counts: inner.mode_attribution_counts.clone(),
        }
    }

    /// Return the current sticky Recall variant's answer mode raw value, or None
    /// when no sticky Recall=<variant> is set.
    ///
    /// Called by the mode concern's pre-decode transform hook in
    /// `aria_v2_pre_decode_registrations` to inject `answer` into `moot_memory_search`
    /// arguments when the per-call `answer` arg is absent.
    pub fn sticky_recall_answer_mode(&self) -> Option<&'static str> {
        let inner = self.inner.lock().expect("ModeSessionState lock poisoned");
        inner.sticky_declaration
            .as_ref()
            .and_then(|d| d.recognized_recall_variant())
            .map(|v| v.answer_mode_raw_value())
    }

    /// Override the coaching cadence. Used by tests to trigger coaching quickly
    /// without provisioning an estate manifest entry. Sets `configured_from_estate`
    /// to true so that `Dispatcher::tools_call` skips the provisioned-config read
    /// on the first tool call — the seam takes full precedence over the estate key.
    ///
    /// Mirrors Swift `ModeSessionState.setCoachingCallsX(_:)`.
    pub fn set_coaching_calls_x(&self, value: usize) {
        let mut inner = self.inner.lock().expect("ModeSessionState lock poisoned");
        inner.coaching_calls_x = value;
        inner.set_configured_from_estate(true);
    }

    /// Apply preferences read from the estate manifest
    /// (`EstateCoordinator::provisioned_modes_config`). Guards itself:
    /// subsequent calls are no-ops (bit 1 = configured_from_estate).
    ///
    /// Called by `Dispatcher::tools_call` on the first tool call of the session.
    /// Mirrors Swift `ModeSessionState.applyPreferences(stickyEnabled:coachingCalls:)`.
    pub fn apply_preferences(&self, sticky_enabled: bool, coaching_calls: usize) {
        let mut inner = self.inner.lock().expect("ModeSessionState lock poisoned");
        if inner.configured_from_estate() {
            return;
        }
        inner.set_sticky_enabled(sticky_enabled);
        inner.coaching_calls_x = coaching_calls;
        inner.set_configured_from_estate(true);
    }

    /// Returns true if `apply_preferences` has already been called this session.
    ///
    /// Guards the apply-once pattern in `Dispatcher::tools_call`.
    pub fn is_configured_from_estate(&self) -> bool {
        let inner = self.inner.lock().expect("ModeSessionState lock poisoned");
        inner.configured_from_estate()
    }
}

impl Default for ModeSessionState {
    fn default() -> Self {
        Self::new()
    }
}
